### PR TITLE
Ifpack2: MT Gauss-Seidel fixup

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/ifpack2 
@trilinos/kokkos-kernels 

## Description
Based on the changes to the MT Gauss-Seidel in kokkos-kernels, this enables damping and the L1 method in Ifpack2's MT GS interface. This deals with the issue that the implementation silently assumed that the matrix diagonal is locally given by indices (i, i), which might not be true for general row and column maps.

This PR includes the changes of https://github.com/kokkos/kokkos-kernels/pull/255, so that they are available before the next promotion of kokkos-kernels (at which point they will be overwritten).

Addresses #2665, #2768

### CODE REVIEW

File: packages/kokkos-kernels/src/sparse/KokkosSparse_gauss_seidel_handle.hpp [hunk 1]
```diff
@@ -50,292 +50,292 @@
 
 namespace KokkosSparse{
 
-enum GSAlgorithm{GS_DEFAULT, GS_PERMUTED, GS_TEAM};
+  enum GSAlgorithm{GS_DEFAULT, GS_PERMUTED, GS_TEAM};
```
Comment (mhoemmen) [inline packages/kokkos-kernels/src/sparse/KokkosSparse_gauss_seidel_handle.hpp]: I'm not a fan of the name of this enum, but it was already there, so you're stuck with it.  Enums in a public namespace should have a human-readable name, like `GaussSeidelAlgorithm`.

File: packages/kokkos-kernels/src/sparse/KokkosSparse_gauss_seidel_handle.hpp [hunk 2]
```diff
@@ -50,292 +50,292 @@
...
-template <class size_type_, class lno_t_, class scalar_t_,
-          class ExecutionSpace,
-          class TemporaryMemorySpace,
-          class PersistentMemorySpace>
-class GaussSeidelHandle{
-public:
-  typedef ExecutionSpace HandleExecSpace;
-  typedef TemporaryMemorySpace HandleTempMemorySpace;
-  typedef PersistentMemorySpace HandlePersistentMemorySpace;
+  template <class size_type_, class lno_t_, class scalar_t_,
```
Comment (mhoemmen) [inline packages/kokkos-kernels/src/sparse/KokkosSparse_gauss_seidel_handle.hpp]: Changing the formatting without changing the content makes reviewing code hard, especially in this case, where you've made a subtle change to the algorithm that needs proofreading.

File: packages/kokkos-kernels/src/sparse/KokkosSparse_gauss_seidel_handle.hpp [hunk 3]
```diff
@@ -50,292 +50,292 @@
...
-   */
-  void choose_default_algorithm(){
+     */
+    void choose_default_algorithm(){
 #if defined( KOKKOS_ENABLE_SERIAL )
-    if (Kokkos::Impl::is_same< Kokkos::Serial , ExecutionSpace >::value){
-      this->algorithm_type = GS_PERMUTED;
+      if (Kokkos::Impl::is_same< Kokkos::Serial , ExecutionSpace >::value){
+        this->algorithm_type = GS_PERMUTED;
 #ifdef VERBOSE
```
Comment (mhoemmen) [inline packages/kokkos-kernels/src/sparse/KokkosSparse_gauss_seidel_handle.hpp]: It's a bad idea to have macros with no namespacing (e.g., `KOKKOSSPARSE_VERBOSE` would be nice), but again, that's not your fault.

File: packages/kokkos-kernels/src/sparse/impl/KokkosSparse_gauss_seidel_impl.hpp [hunk 4]
```diff
@@ -55,1545 +55,1608 @@
 namespace KokkosSparse{
 
 
-namespace Impl{
+  namespace Impl{
```
Comment (mhoemmen) [inline packages/kokkos-kernels/src/sparse/impl/KokkosSparse_gauss_seidel_impl.hpp]: See my comment above -- unnecessary formatting changes make code reviews hard.

**Comment by cgcgcg** [comment]
@mhoemmen Thank you for reviewing so quickly! I'll go through them tomorrow.

Unfortunately, the kokkkos-kernels files where formatted with a mix of tabs and whitespaces, so I decided to blow all of that away. And given that this was going to be a huge diff already, I figured that I might as well have the whole thing indented identically..

What kind of test where you thinking of? I can obviously add something to packages/ifpack2/test/belos/ that exercises the different MT GS options..

**Comment by mhoemmen** [comment]
@cgcgcg wrote:
> Unfortunately, the kokkkos-kernels files where formatted with a mix of tabs and whitespaces, so I decided to blow all of that away. And given that this was going to be a huge diff already, I figured that I might as well have the whole thing indented identically.

Ah, I see, sorry about the complaint then :) 

> What kind of test were you thinking of? I can obviously add something to packages/ifpack2/test/belos/ that exercises the different MT GS options.

It might be nice to have a test that exercises both standard Gauss-Seidel and this MT version with a nondefault omega, and/or a nondefault diagonal.  It doesn't matter if Belos converges; what matters is that the `apply()` method does the same thing when the thread count is 1.

File: packages/ifpack2/src/Ifpack2_Relaxation_def.hpp [hunk 5]
```diff
@@ -616,58 +616,62 @@ void Relaxation<MatrixType>::initialize ()
...
-      std::stringstream ss;
-      ss << random_integer;
-      std::string str = ss.str();
-      Tpetra::MatrixMarket::Writer<crs_matrix_type> crs_writer;
-      std::string file_name = str + "_Ifpack2_MT_GS.mtx";
-      Teuchos::RCP<const crs_matrix_type> rcp_crs_mat = Teuchos::rcp_dynamic_cast<const crs_matrix_type> (A_);
-      crs_writer.writeSparseFile(file_name, rcp_crs_mat);
-    }
+    if (PrecType_ == Ifpack2::Details::MTGS || PrecType_ == Ifpack2::Details::MTSGS) {
+      const crs_matrix_type* crsMat = dynamic_cast<const crs_matrix_type*> (&(*A_));
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/ifpack2/src/Ifpack2_Relaxation_def.hpp]: Done!

File: packages/ifpack2/src/Ifpack2_Relaxation_def.hpp [hunk 6]
```diff
@@ -616,58 +616,62 @@ void Relaxation<MatrixType>::initialize ()
...
+      TEUCHOS_TEST_FOR_EXCEPTION
+        (crsMat == NULL, std::logic_error, "Ifpack2::Relaxation::initialize: "
+         "Multithreaded Gauss-Seidel methods currently only work when the input "
+         "matrix is a Tpetra::CrsMatrix.");
 
-    this->mtKernelHandle_ = Teuchos::rcp (new mt_kernel_handle_type ());
-    if (mtKernelHandle_->get_gs_handle () == NULL) {
-      mtKernelHandle_->create_gs_handle ();
+      if(this->ifpack2_dump_matrix_){
+        int random_integer = rand();
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/ifpack2/src/Ifpack2_Relaxation_def.hpp]: Changed to use the PID instead.

**Comment by cgcgcg** [comment]
@mhoemmen Added some tests and realized that I hard coded the damping factor :-/
The iteration counts between classical and MT Gauss-Seidel do not match, probably because of the different ordering of unknowns. (But they are of the same order of magnitude.)

File: packages/ifpack2/src/Ifpack2_Relaxation_def.hpp [hunk 7]
```diff
@@ -52,6 +52,7 @@
 #include "MatrixMarket_Tpetra.hpp"
 #include <cstdlib>
 #include <sstream>
+#include <unistd.h>
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Relaxation_def.hpp]: Please see comment below.  I know our Windows build situation isn't so good for the Tpetra stack right now, but I wouldn't like to make it worse.

File: packages/ifpack2/src/Ifpack2_Relaxation_def.hpp [hunk 8]
```diff
@@ -634,19 +635,15 @@ void Relaxation<MatrixType>::initialize ()
...
         (crsMat == NULL, std::logic_error, "Ifpack2::Relaxation::initialize: "
          "Multithreaded Gauss-Seidel methods currently only work when the input "
          "matrix is a Tpetra::CrsMatrix.");
 
       if(this->ifpack2_dump_matrix_){
-        int random_integer = rand();
-        std::stringstream ss;
-        ss << random_integer;
-        std::string str = ss.str();
+        std::string file_name = "Ifpack2_MT_GS." + std::to_string(::getpid()) + ".mtx";
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Relaxation_def.hpp]: This is both wrong, and ineffective.

  1. Different MPI processes may (or may not!) have different OS process IDs.
  2. Tpetra::MatrixMarket::Writer is a collective, and it writes a single file.  All processes in the matrix's communicator need to agree on the filename.

Do this instead:
```c++
if (this->ifpack2_dump_matrix_) {
  static int sequence_number = 0;
  const std::string file_name = "Ifpack2_MT_GS_" + std::to_string (sequence_number++) + ".mtx";
  // ... etc. ...
}
```
Also, we really don't want to use POSIX stuff not in the C++ standard unless there's no other way.

**Comment by cgcgcg** [comment]
@mhoemmen Ah, dang. It was stuck in my mind that I was trying to dump local matrices..

**Comment by mhoemmen** [comment]
No worries! :D 

@cgcgcg wrote:
> It was stuck in my mind that I was trying to dump local matrices.

Even then, you would want the MPI process rank, not the OS process ID.  Suppose that different processes run on different nodes, but access a shared file system.  Different nodes run different OS instances, and thus might have the same OS process ID, despite being different processes.

**Comment by cgcgcg** [comment]
@trilinos/framework What's happening here? Why is the autotester running 6 builds?

**Comment by jwillenbring** [comment]
@cgcgcg The first set of builds failed due to a GitHub communication problem all PR testing is having. @allevin is looking into it, but at first glance it just appears to be slow/timing out communication with GitHub.

**Comment by prwolfe** [comment]
@cgcgcg It looks like the python code that's supposed to monitor the builds had a timeout while asking GitHub if the PR still exists.  It looks like it did not kill the existing jobs, and the next time it started (runs every 20 minutes) it reset and restarted the jobs.  I will talk with @allevin about how we could better notify you and stop the existing jobs when something like this happens.

I expect you will only get reports about the status of the second set of jobs here, but both should report to CDash, just to make it even more confusing.

**Comment by cgcgcg** [comment]
I just wanted to make sure the autotester wasn't running wild and spawning new, additional builds at every single iteration :-)

**Comment by allevin** [comment]
@cgcgcg @trilinos/framework Basically it looks like we had a 30-45 minute (starting around noon) hiccup on comms with Github (not sure if it is our side or Github's).  

The autotester will retry any failed communication up to 3 times with a 20 second delay in between each attempt.  Once the 3rd attempt fails, it basically throws up its hands and gives up.  Unfortunately, since it assumes Github comms are down, there is no way to notify the user of this situation (The only way would be to post a comment into the Github PR comments... but Github comms are down).  

When the autotester gives up, the error handler does not know that there are existing jobs running; therefore, those jobs are orphaned.  Since we have very little state info for the Autotester between runs, it just assumes that there was a problem and starts testing again.  The messages can be a bit confusing when this happens.  

We do have an open issue on this, but its actually harder than one would think to resolve.  I will take a new look at this and see if I can get some sort of message to be posted when the autotester re-starts under this situation.

**Comment by aprokop** [comment]
Can this be merged? I'd like to try to run Nalu with it on Cori, and preferably I need a stable hash from `@develop`. Or is this waiting to address @srajama1's comments in Kokkos?

**Comment by cgcgcg** [comment]
I wanted to wait for the kokkos-kernels PR to be merged first. I also am not convinced anymore that what I did is the perfect solution. (It should work better than what we had, but there might still be some issues.)

**Comment by cgcgcg** [comment]
@jhux2 Thanks!
